File: backend/tools/internet_search/fetch_webpage.py
import logging
import requests
import re
import urllib3

# Disable warnings about SSL verification
urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)

logger = logging.getLogger(__name__)

TOOL_DEF = {
    "name": "fetch_webpage",
    "description": "Fetches and extracts the main content from a webpage URL. Returns the title, full text content, and metadata.",
    "input_schema": {
        "type": "object",
        "properties": {
            "url": {
                "type": "string",
                "description": "The URL of the webpage to fetch.",
            },
            "extract_main_content": {
                "type": "boolean",
                "description": "Whether to extract just the main content (true) or return the full HTML (false).",
                "default": True,
            },
        },
        "required": ["url"],
    },
    "default_enabled": True,
}

def extract_title(html_content):
    """Extract title from HTML using regex."""
    title_match = re.search(r'<title[^>]*>(.*?)</title>', html_content, re.IGNORECASE | re.DOTALL)
    if title_match:
        return title_match.group(1).strip()
    return ""

def simple_html_to_text(html_content):
    """Very simple HTML to text conversion using regex."""
    # Remove script and style elements
    html_content = re.sub(r'<script[^>]*>.*?</script>', '', html_content, flags=re.DOTALL | re.IGNORECASE)
    html_content = re.sub(r'<style[^>]*>.*?</style>', '', html_content, flags=re.DOTALL | re.IGNORECASE)
    
    # Replace common block elements with newlines
    html_content = re.sub(r'</(div|p|h1|h2|h3|h4|h5|h6|article|section|main)>', '\n\n', html_content, flags=re.IGNORECASE)
    
    # Remove all other HTML tags
    html_content = re.sub(r'<[^>]*>', '', html_content)
    
    # Clean up whitespace
    html_content = re.sub(r'\n\s*\n', '\n\n', html_content)
    html_content = re.sub(r'[ \t]+', ' ', html_content)
    
    return html_content.strip()

def call_tool(tool_input: dict, context=None):
    url = (tool_input.get("url") or "").strip()
    extract_content = tool_input.get("extract_main_content", True)
    
    if not url:
        return 400, "url is required and cannot be empty. Provide a valid webpage URL."
    
    try:
        # Set a user agent to avoid being blocked
        headers = {
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36"
        }
        
        # Disable SSL verification to handle certificate issues
        response = requests.get(url, headers=headers, timeout=15, verify=False)
        
        # Check if the request was successful
        if response.status_code != 200:
            return 500, f"Error: HTTP status code {response.status_code}"
            
        html_content = response.text
        
        # Extract the title
        title = extract_title(html_content)
        
        # Get metadata
        metadata = {
            "url": url,
            "status_code": response.status_code,
            "content_type": response.headers.get('Content-Type', ''),
        }
        
        # Extract content based on user preference
        if extract_content:
            content = simple_html_to_text(html_content)
        else:
            content = html_content
        
        logger.info(f"Successfully fetched webpage: {url}")
        return 200, {
            "title": title,
            "content": content,
            "metadata": metadata,
        }
    
    except requests.exceptions.RequestException as e:
        logger.error(f"Error fetching webpage {url}: {str(e)}")
        return 500, f"Error fetching webpage: {str(e)}"
    except Exception as e:
        logger.error(f"Unexpected error processing webpage {url}: {str(e)}")
        return 500, f"Unexpected error: {str(e)}"

File: akira-desktop/electron/tools/web/fetch-webpage.js
/**
 * Fetch Webpage Tool
 * fetch_webpage - Fetch and extract content from a webpage URL
 */

const https = require('https');
const http = require('http');
const { URL } = require('url');
const zlib = require('zlib');

/**
 * Simple HTML to text conversion
 */
function htmlToText(html) {
  // Remove script and style elements
  let text = html.replace(/<script[^>]*>[\s\S]*?<\/script>/gi, '');
  text = text.replace(/<style[^>]*>[\s\S]*?<\/style>/gi, '');

  // Replace block elements with newlines
  text = text.replace(/<\/(div|p|h1|h2|h3|h4|h5|h6|article|section|main|br)>/gi, '\n');

  // Remove all other HTML tags
  text = text.replace(/<[^>]*>/g, '');

  // Decode HTML entities
  text = text.replace(/&nbsp;/g, ' ');
  text = text.replace(/&amp;/g, '&');
  text = text.replace(/&lt;/g, '<');
  text = text.replace(/&gt;/g, '>');
  text = text.replace(/&quot;/g, '"');
  text = text.replace(/&#39;/g, "'");

  // Clean up whitespace
  text = text.replace(/\n\s*\n/g, '\n\n');
  text = text.replace(/[ \t]+/g, ' ');

  return text.trim();
}

/**
 * Extract title from HTML
 */
function extractTitle(html) {
  const match = html.match(/<title[^>]*>([\s\S]*?)<\/title>/i);
  return match ? match[1].trim() : '';
}

/**
 * Make HTTP/HTTPS request
 */
function fetchUrl(url, options = {}) {
  return new Promise((resolve, reject) => {
    const parsed = new URL(url);
    const protocol = parsed.protocol === 'https:' ? https : http;

    const reqOptions = {
      hostname: parsed.hostname,
      port: parsed.port,
      path: parsed.pathname + parsed.search,
      method: options.method || 'GET',
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36',
        'Accept-Encoding': 'gzip, deflate',
        ...options.headers,
      },
      timeout: options.timeout || 15000,
      rejectUnauthorized: false, // Allow self-signed certs
    };

    const req = protocol.request(reqOptions, (res) => {
      const chunks = [];

      // Handle redirects
      if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
        const redirectUrl = new URL(res.headers.location, url).toString();
        fetchUrl(redirectUrl, options).then(resolve).catch(reject);
        return;
      }

      res.on('data', (chunk) => {
        chunks.push(chunk);
      });
      res.on('end', () => {
        const buffer = Buffer.concat(chunks);
        const encoding = res.headers['content-encoding'];

        // Decompress if needed
        let data;
        try {
          if (encoding === 'gzip') {
            data = zlib.gunzipSync(buffer).toString('utf8');
          } else if (encoding === 'deflate') {
            data = zlib.inflateSync(buffer).toString('utf8');
          } else {
            data = buffer.toString('utf8');
          }
        } catch (err) {
          // If decompression fails, try raw buffer
          data = buffer.toString('utf8');
        }

        resolve({
          statusCode: res.statusCode,
          headers: res.headers,
          data,
        });
      });
    });

    req.on('error', reject);
    req.on('timeout', () => {
      req.destroy();
      reject(new Error('Request timed out'));
    });

    req.end();
  });
}

const definitions = [
  {
    name: 'fetch_webpage',
    description: 'Fetch and extract content from a webpage URL. Returns title and text content.',
    input_schema: {
      type: 'object',
      properties: {
        url: {
          type: 'string',
          description: 'The URL of the webpage to fetch',
        },
        extract_main_content: {
          type: 'boolean',
          description: 'Extract just text content (true) or return full HTML (false). Default: true',
        },
        max_length: {
          type: 'integer',
          description: 'Maximum content length to return (default: 50000)',
        },
      },
      required: ['url'],
    },
  },
];

const handlers = {
  async fetch_webpage(input) {
    const url = (input.url || '').trim();
    const extractContent = input.extract_main_content !== false;
    const maxLength = input.max_length || 50000;

    if (!url) {
      return { success: false, error: 'URL is required' };
    }

    // Validate URL
    try {
      new URL(url);
    } catch {
      return { success: false, error: 'Invalid URL format' };
    }

    try {
      const response = await fetchUrl(url, { timeout: 15000 });

      if (response.statusCode !== 200) {
        return { success: false, error: `HTTP status ${response.statusCode}`, url };
      }

      const html = response.data;
      const title = extractTitle(html);

      let content;
      if (extractContent) {
        content = htmlToText(html);
        if (content.length > maxLength) {
          content = content.substring(0, maxLength) + '\n\n[Content truncated...]';
        }
      } else {
        content = html.substring(0, maxLength);
      }

      return {
        success: true,
        title,
        content,
        url,
        content_type: response.headers['content-type'] || '',
      };
    } catch (error) {
      return { success: false, error: error.message, url };
    }
  },
};

module.exports = { definitions, handlers };
